### ISSUE TITLE
[BUG] App blocks nvidia advanced optimus (automatic switch between gpu)

### ISSUE THREAD

**Issue by dopealicious** [unchihugo/FluentFlyout#15]
**Describe the bug**
App itself blocks nvidia advanced optimus (automatic switch between integrated gpu and discrete gpu). I think it has something to do with dx9, but I could be wrong.

**To Reproduce**
Just try to switch gpu while app is running.

**Desktop (please complete the following information):**
 - OS: Windows 11 23H2
 - Version 1.3.0.0
![`block](https://github.com/user-attachments/assets/efcc047a-b998-4f29-b063-7939357b590c)

**Comment by unchihugo** [comment]
Hi! While I can't reproduce this issue for myself (switching display mode to Optimus and vice versa works on my laptop), I have looked around and stumbled upon this issue in the official WPF repo: https://github.com/dotnet/wpf/issues/9286. It seems like it's an issue with Nvidia?

**Comment by himischa** [comment]
I don't know if I can answer that
(+according to link that you gave) it happened too with steamwebhelper.exe back with game drivers prior to 551.76 (but it didn't happen **before** Steam did something that made Nvidia have to fix it in their driver update). I'm not trying to blame you (awesome app, I like it) but it "unpleasant" to close the app everytime the system try to switch gpu.

**Comment by unchihugo** [comment]
I totally understand that!

Maybe setting FluentFlyout.exe to integrated graphics in the Nvidia Control Panel, under Manage 3D Settings > Program Settings would help fix this?

**Comment by himischa** [comment]
nope, the problem still occurs. I forced it in both windows settings and nvcp.

**Comment by binsentlangit** [comment]
Hello, I also have this problem. An app that also had this issue was DS4Windows. Sadly, I don't really know how NVIDIA's advanced optimus works so there's not much help I can give here.

**Comment by lolmam** [comment]
04-2025 - the problem is still occuring, with FluentFlyout and DS4Windows - using latest nvidia drivers

**Comment by HoshinekoSanti** [comment]
06-2026 - Issue still persists. Tried setting FluentFlyout with integrated graphics but it doesn't resolve the issue.

Using NVIDIA GeForce RTX 5050 Laptop GPU - Driver version 581.95 - DirectX 12.0

**Comment by Zar-ar** [comment]
I was just looking for a post that fixes this issue for DS4Windows and Warp, to fix the same issue with DSX.

Stumbled here, and while i cant be sure since i dont know what the project is for at all, this might fix it for you as well.
https://www.reddit.com/r/GamingLaptops/comments/1aq305j/a_workaround_to_apps_that_prevent_advanced/

Its because advanced optimus doesnt work too well with DX9. If FluentFlyout doesnt use that, then please disregard.

**Comment by jin6yang** [comment]
### September 2026 update: The issue is still reproducible and is accompanied by abnormally high CPU package power at idle

I can confirm that the issue described here is still present in the current version. When FluentFlyout is running in the background, NVIDIA App explicitly identifies it as an application that is blocking an Advanced Optimus/DDS display-mode switch.

On my device, the issue not only prevents switching back to **Automatic Select**, but is also accompanied by an abnormal hybrid-graphics state, persistent activity from the `System` process, and unusually high CPU package power and temperature while the desktop is idle.

### System information

- Laptop model: ASUS GA403WP
- CPU: AMD Ryzen AI 9 HX 370
- Integrated GPU: AMD Radeon 890M
- Discrete GPU: NVIDIA GeForce RTX 5070 Laptop GPU
- GPU configuration: ASUS Armoury Crate Hybrid Output (MSHybrid) with NVIDIA Advanced Optimus/DDS
- Windows: Windows 11 25H2, OS Build 26200.9168
- NVIDIA driver version: 596.36
- Armoury Crate version: 6.5.7.0
- FluentFlyout version: 2.14.0
- Installation source: Microsoft Store

The laptop was connected to AC power throughout the test. The power mode, background applications, and other test conditions remained unchanged. The AMD NPU was disabled during both measurements.

“Idle” in this report means an idle Windows desktop, not sleep or Modern Standby.

### Steps to reproduce

1. Set Armoury Crate to **Hybrid Output / MSHybrid** mode.
2. Start FluentFlyout and leave it running in the background.
3. In NVIDIA App, switch the laptop display mode to **NVIDIA GPU**.
4. Attempt to switch it back to **Automatic Select**.
5. NVIDIA App reports that an application is preventing the switch and explicitly lists FluentFlyout.
6. The switch fails, and the system cannot return to Automatic Select mode normally.

The issue appears to occur more readily after media playback or after triggering the volume, media, or taskbar flyout, although this has not yet been fully verified.

### Behavior in the affected state

While the hybrid-graphics state is affected:

- NVIDIA Advanced Optimus/DDS cannot complete the display-mode switch.
- Armoury Crate and NVIDIA App may report inconsistent GPU modes.
- CPU Package Power remains at approximately 20–30 W while the desktop is idle or under light load.
- CPU temperature remains at approximately 70–85°C.
- Windows Task Manager normally reports around 12% total CPU utilization, generally remaining below 20%.
- Armoury Crate reports approximately 25% CPU utilization, with occasional spikes to around 60%.
- The `System` process shows persistent CPU activity.
- The discrete GPU remains awake despite GPU utilization being close to 0%. AIDA64 reports approximately 12 W of non-zero GPU power.
- The following entries repeatedly appear among the active `System` threads in Process Explorer:
  - `dxgmms2.sys!VidMmInterface+0x94ed0`
  - Multiple kernel worker threads whose start addresses are displayed as `ntoskrnl.exe!KeSetBasePriorityThread+0x5750`

These are the thread start addresses displayed by Process Explorer, not complete live call stacks. They therefore appear to be symptoms of the issue and do not, by themselves, identify the root cause.

### Control test

With all other test conditions unchanged, I switched the display mode to **NVIDIA GPU / discrete-GPU-only** and waited approximately 10 minutes.

After doing so:

- The abnormal CPU activity largely disappeared.
- CPU Package Power decreased from approximately 20–30 W to approximately 7 W.
- CPU temperature and `System` process utilization decreased accordingly.
- Activity associated with `dxgmms2.sys` was significantly reduced.

The approximately 7 W measurement refers only to CPU Package Power and does not include the discrete GPU’s own power consumption.

This result suggests that the behavior is unlikely to be caused by an ordinary background CPU workload. It appears to be strongly associated with the hybrid-graphics path, graphics-resource residency, or a failed DDS transition.

### Expected behavior

- FluentFlyout should not prevent NVIDIA Advanced Optimus/DDS transitions.
- The system should be able to switch normally between **NVIDIA GPU**, **Automatic Select**, and **Optimus**.
- FluentFlyout should release graphics devices, swap chains, or composition resources that are no longer required after a flyout is hidden.
- FluentFlyout should not keep the discrete GPU awake when there is no active media interaction.
- A failed display-mode transition should not be accompanied by persistent `System` or `dxgmms2.sys` activity or abnormally high CPU Package Power.

### Initial hypothesis

Based on the observed behavior, FluentFlyout or one of its UI/graphics components may be retaining a Direct3D, DXGI, DirectComposition, WPF/Mica, or swap-chain resource that is bound to a specific GPU adapter.

This could cause the NVIDIA driver to treat FluentFlyout as a GPU client that cannot be safely migrated or released, preventing DDS from completing the display-mode transition.

This is only a hypothesis based on the observed behavior and has not been confirmed through a complete call stack or source-level debugging.

Potential areas to investigate include:

1. Whether FluentFlyout explicitly or indirectly selects the high-performance GPU.
2. Whether graphics resources bound to the discrete GPU remain active after a flyout or taskbar component is hidden.
3. Whether the application correctly handles display-adapter changes, device loss, and display-configuration changes.
4. Whether FluentFlyout could request the power-saving GPU by default or provide an option to force integrated graphics or disable hardware acceleration.

### Attachments
(Apologies that the screenshots are in Chinese.)

- NVIDIA App screenshot showing FluentFlyout blocking Automatic Select

<img width="2272" height="1649" alt="Image" src="https://github.com/user-attachments/assets/5db527dd-c4b2-433e-a4ed-18c9d9ec8af0" />

- Armoury Crate and NVIDIA App GPU-mode status screenshots

<img width="2880" height="1716" alt="Image" src="https://github.com/user-attachments/assets/61939eda-b9b9-4519-8c7d-f4585aeceb35" />

<img width="2272" height="1649" alt="Image" src="https://github.com/user-attachments/assets/fed5d2ba-dc8d-47c2-b495-883d884220d5" />

<img width="2880" height="1716" alt="Image" src="https://github.com/user-attachments/assets/e7184824-600e-466c-9b4a-67d02b291230" />

- CPU Package Power comparison before and after switching modes

Before:
<img width="852" height="1182" alt="Image" src="https://github.com/user-attachments/assets/ad39e8e7-3007-4b4f-9af3-581eae46a608" />

After:
<img width="761" height="1163" alt="Image" src="https://github.com/user-attachments/assets/c2c945ea-c293-49cc-93f3-f119c7aa30f3" />

- Process Explorer screenshots showing the `System` and `dxgmms2.sys` threads

<img width="1226" height="1444" alt="Image" src="https://github.com/user-attachments/assets/43977e68-e86c-4926-bc68-e3e32d495888" />

### Possible additional verification

The following checks could help isolate the behavior further:

1. Check whether fully exiting FluentFlyout immediately restores Advanced Optimus switching.
2. In Automatic/Optimus mode, restart FluentFlyout and trigger a media or volume flyout to determine whether the issue can be reproduced consistently.
3. Force FluentFlyout to use **Power saving / AMD Radeon 890M** in Windows Graphics settings and check whether the issue still occurs.
